File: 0057-Insert-Interval.js
var insert = function(intervals, newInterval) {
    let [leftEnd, rightEnd] = newInterval;
    let leftCollection = [];
    let rightCollection = [];

    for (const interval of intervals) {
        const [currentLeft, currentRight] = interval;

        if (currentRight < leftEnd) {
            leftCollection.push(interval);
        } else if (currentLeft > rightEnd) {
            rightCollection.push(interval);
        } else {
            leftEnd = Math.min(leftEnd, currentLeft);
            rightEnd = Math.max(rightEnd, currentRight);
        }
    }

    return [...leftCollection, [leftEnd, rightEnd], ...rightCollection];
};

File: 0057-Insert-Interval.py
class Solution(object):
    def insert(self, intervals, newInterval):
        n = len(intervals)
        if n == 0:
            return [newInterval]

        leftEnd, rightEnd = list(map(list, zip(*intervals)))
        newLeft, newRight = newInterval

        leftIns = bisect_right(leftEnd, newLeft)
        rightIns = bisect_left(rightEnd, newRight)

        if leftIns > 0 and newLeft <= rightEnd[leftIns-1]:
            newLeft = leftEnd[leftIns - 1]
            leftIns -= 1

        if rightIns < n and newRight >= leftEnd[rightIns]:
            newRight = rightEnd[rightIns]
            rightIns += 1

        return intervals[:leftIns] + [[newLeft, newRight]] + intervals[rightIns:]
